File: Homework/quoc_khanh/lesson_2/20.ValidParentheses.py
from queue import Empty
from typing import List

TOP = -1
EMPTY = ""
openToClose = {
    '{': '}',
    '[': ']',
    '(': ')'
}

def isOpenBracket(bracket: str) -> bool:
    return openToClose.get(bracket, EMPTY) != EMPTY

def isEmpty(stack: List[str]) -> bool:
    return len(stack) == 0

def top(stack: List[str]) -> str:
    if isEmpty(stack):
        return EMPTY
    else:
        return stack[TOP]

def isMatched(openBracket: str, closeBracket: str) -> bool:
    return openToClose.get(openBracket, EMPTY) == closeBracket

class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for bracket in s:
            if isOpenBracket(bracket):
                stack.append(bracket)
            else:
                if isMatched(top(stack), bracket):
                    stack.pop()
                else:
                    return False
        return len(stack) == 0

s = Solution()
print(s.isValid('{}[]()') == True)
print(s.isValid('([{}])') == True)
print(s.isValid('({[}])') == False)
print(s.isValid('{') == False)
print(s.isValid(']') == False)
print(s.isValid('[]') == True)

File: Homework/Duc_Trinh/lesson-3/20-valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
 var isValid = function(s) {
    const bracketsMap = {
        '{': '}',
        '[': ']',
        '(': ')',
    }
    let stack = []
    for (let i = 0; i < s.length; i++) {
        if (Object.keys(bracketsMap).indexOf(s[i]) > -1) {
            stack.push(s[i])
        } else {
            const top = stack[stack.length - 1]
            if (s[i] === bracketsMap[top]) {
                stack.pop()
            } else {
                return false
            }
        }
    }
    return stack.length === 0
  };